alpine/FreeElectronLaser.cpp

```diff
@@ -0,0 +1,1851 @@
+        gamma_m = sqrt(1 + gammaBeta * gammaBeta);
+    }
+    KOKKOS_INLINE_FUNCTION void primedToUnprimed(Vector3& arg, scalar time)const noexcept{
+        arg[axis] = gamma_m * (arg[axis] + beta_m * time); 
+    }
+    KOKKOS_INLINE_FUNCTION Kokkos::pair<ippl::Vector<T, 3>, ippl::Vector<T, 3>> transform_EB(const Kokkos::pair<ippl::Vector<T, 3>, ippl::Vector<T, 3>>& unprimedEB)const noexcept{
+        
+        Kokkos::pair<ippl::Vector<T, 3>, ippl::Vector<T, 3>> ret;
+        ippl::Vector<scalar, 3> betavec{0, 0, beta_m};
+        ret.first  = ippl::Vector<T, 3>(unprimedEB.first  + betavec.cross(unprimedEB.second)) * gamma_m;// - (vnorm * (gamma_m - 1) * (unprimedEB.first.dot(vnorm)));
+        ret.second = ippl::Vector<T, 3>(unprimedEB.second - betavec.cross(unprimedEB.first )) * gamma_m;// - (vnorm * (gamma_m - 1) * (unprimedEB.second.dot(vnorm)));
+        ret.first[axis] -= (gamma_m - 1) * unprimedEB.first[axis];
+        ret.second[axis] -= (gamma_m - 1) * unprimedEB.second[axis];
+        return ret;
+    }
+    KOKKOS_INLINE_FUNCTION Kokkos::pair<ippl::Vector<T, 3>, ippl::Vector<T, 3>> inverse_transform_EB(const Kokkos::pair<ippl::Vector<T, 3>, ippl::Vector<T, 3>>& primedEB)const noexcept{
+        return negative().transform_EB(primedEB);
+    }
+};
+template<typename scalar>
+BunchInitialize<scalar> generate_mithra_config(const config& cfg, const UniaxialLorentzframe<scalar>& /*frame_boost unused*/) {
+    using vec3 = ippl::Vector<scalar, 3>;
+    scalar frame_gamma = cfg.bunch_gamma / std::sqrt(1 + 0.5 * cfg.undulator_K * cfg.undulator_K);
+    BunchInitialize<scalar> init;
+    init.generator_ = "random";
+    init.distribution_ = "uniform";
+    init.initialDirection_ = vec3{0, 0, 1};
+    init.initialGamma_ = cfg.bunch_gamma;
+    init.initialBeta_ = cfg.bunch_gamma == scalar(1) ? 0 : (sqrt(cfg.bunch_gamma * cfg.bunch_gamma - 1) / cfg.bunch_gamma);
+    init.sigmaGammaBeta_ = cfg.sigma_momentum.template cast<scalar>();
+    init.sigmaPosition_ = cfg.sigma_position.template cast<scalar>();
+
+    // TODO: Initial bunching factor huh
+    init.bF_ = 0.01;
+    init.bFP_ = 0;
+    init.shotNoise_ = false;
+    init.cloudCharge_ = cfg.charge;
+    init.lambda_ = cfg.undulator_period / (2 * frame_gamma * frame_gamma);
+    init.longTrun_ = cfg.position_truncations[2];
+    init.tranTrun_ = cfg.position_truncations[0];
+    init.position_ = cfg.mean_position.cast<scalar>();
+    init.betaVector_ = ippl::Vector<scalar, 3>{0,0,init.initialBeta_};
+    init.numberOfParticles_ = cfg.num_particles;
+
+    init.numbers_ = 0;                      // UNUSED
+    init.latticeConstants_ = vec3{0, 0, 0}; // UNUSED
+
+    return init;
+}
+template <typename Double>
+struct Charge {
+
+    Double q;                     /* Charge of the point in the unit of electron charge.	*/
+    FieldVector<Double> rnp, rnm; /* Position vector of the charge.			*/
+    FieldVector<Double> gb;       /* Normalized velocity vector of the charge.		*/
+
```

Comment:
Why do you need such a struct? Isn't the particle bunch already containing q, r and v?